flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java

```diff
@@ -206,27 +207,35 @@ public boolean processInput() throws Exception {
 				}
 			}
 
-			final BufferOrEvent bufferOrEvent = barrierHandler.getNextNonBlocked();
-			if (bufferOrEvent != null) {
-				if (bufferOrEvent.isBuffer()) {
-					currentChannel = bufferOrEvent.getChannelIndex();
-					currentRecordDeserializer = recordDeserializers[currentChannel];
-					currentRecordDeserializer.setNextBuffer(bufferOrEvent.getBuffer());
-				}
-				else {
-					// Event received
-					final AbstractEvent event = bufferOrEvent.getEvent();
-					if (event.getClass() != EndOfPartitionEvent.class) {
-						throw new IOException("Unexpected event: " + event);
+			final Optional<BufferOrEvent> bufferOrEvent = barrierHandler.pollNext();
+			if (bufferOrEvent.isPresent()) {
+				processBufferOrEvent(bufferOrEvent.get());
+			} else {
+				if (!barrierHandler.isFinished()) {
+					barrierHandler.isAvailable().get();
```

Comment:
This whole method is bit of a mess for my taste. Still wonder if we would not rewrite this part a little so that `processBufferOrEvent` is still called in the same iteration when availability is signaled and not after another roundtrip?

Reply:
Does this means we will waste `the time of waiting for new data` delaying execute mail for the this roundtrip?

Reply:
@Aitozi at the moment yes. To make it work correctly with mailbox thread, this change of non blocking `BarrierHandler` has to be integrated with pending @StefanRRichter 's PR.